### PR TITLE
feat: support `densities` for devices with `DevicePixelRatio` > 1

### LINKED ISSUE #216: Responsive image sizing
We live in a time when websites should work on a huge range of devices with different sizes and densities, it is not enough to have only one version of image... at least 2 versions of each file (1x and 2x) is a must for modern web, but in many cases it may be reasonable to have 3-4 sizes of each file in webp format + fallback to jpg/png, so each picture usually may be resized/converted to 2-4-6 or more versions.

Problems:
- There is no way do define image sizes for high density image versions
- Presets do not allow to use `sizes` property available for `nuxt-img`
-  (for @nuxt/content) Images inside .md files that included using .md syntax are not processed by @nuxt/image, it would be very nice if it is possible to set a default sizing using preset for those images

### PR DESCRIPTION
**Pull request by Mihanik71**
Add densities param for issues #216 

Add:

- default value for densities in nuxt.config
- add param for img and picture components

resolves #216
closes https://github.com/nuxt/image/pull/803

### CODE REVIEW

**Comment by simonkuran** [comment]
I could really use this feature. Would love to see this merged and released!

**Comment by Mihanik71** [comment]
> I could really use this feature. Would love to see this merged and released!

Спасибо! Исправил конфликт для того что бы можно было легко merge

**Comment by Mihanik71** [comment]
> I could really use this feature. Would love to see this merged and released!

Thank you! Fixed the conflict so that you can easily merge

**Comment by Kolobok12309** [comment]
Big thanks for this pr, i write something similar #803, but your solution more cleanest
Found some small bugs

1. Not worked for `sizes` images, fixed by adding `densities: props.densities` here
https://github.com/nuxt/image/blob/381b9063976af1c9e1198776826e0ea1409e8bfa/src/runtime/components/_base.ts#L68-L73

2. `srcset` entries may duplicate, for example in this case, 400
```vue
<nuxt-img
  src="/logos/nuxt.png"
  sizes="sm:200px lg:400px"
  densities="1x 2x 3x"
/>
```
```
/_ipx/w_200/logos/nuxt.png 200w, /_ipx/w_400/logos/nuxt.png 400w, /_ipx/w_400/logos/nuxt.png 400w, /_ipx/w_600/logos/nuxt.png 600w, /_ipx/w_800/logos/nuxt.png 800w, /_ipx/w_1200/logos/nuxt.png 1200w
```

**Review by danielroe — COMMENTED** [review]
Would you be able to have a look at the failing tests?

I also think we can default to `[1, 2]` as default densities.

File: docs/content/5.configuration.md [hunk 1]
```diff
@@ -154,6 +154,20 @@ export default defineNuxtConfig({
 })
 ```
 
+## `densities`
+
+Default: `[]`
```
Comment (hartmut-co-uk) [inline docs/content/5.configuration.md:159]:
```suggestion
Default: `[1, 2]`
```

File: docs/content/5.configuration.md [hunk 2]
```diff
@@ -154,6 +154,20 @@ export default defineNuxtConfig({
...
+
+Specify a value to work with DevicePixelRation > 1 (these are devices with Retina display and others). You must specify for which DevicePixelRation values you want to adapt images.
+
+```ts [nuxt.config.ts]
+export default defineNuxtConfig({
+  image: {
+      densities: [1, 2],
+  }
+})
```
Comment (hartmut-co-uk) [inline docs/content/5.configuration.md:168]:
```suggestion
export default defineNuxtConfig({
  image: {
    densities: [1, 2, 3],
  }
})
```

File: src/runtime/components/nuxt-img.ts [hunk 3]
```diff
@@ -41,6 +41,16 @@ export default defineComponent({
       if (props.sizes) {
         attrs.sizes = sizes.value.sizes
         attrs.srcset = sizes.value.srcset
+      } else if (props.densities || $img.options.densities) {
+        attrs.srcset = $img.getDensitySet(props.src, {
```
Comment (hartmut-co-uk) [inline src/runtime/components/nuxt-img.ts:45]:
```suggestion
        attrs.srcset = $img.getDensitySet(props.src!, {
```

File: src/runtime/utils/index.ts [hunk 4]
```diff
@@ -91,3 +91,11 @@ export function parseSize (input: string | number | undefined = '') {
     }
   }
 }
+
+export function parseDensities (input: string | undefined = '') :number[] {
```
Comment (hartmut-co-uk) [inline src/runtime/utils/index.ts:95]:
```suggestion
export function parseDensities (input: string | undefined = ''): number[] {
```

File: src/runtime/image.ts [hunk 5]
```diff
@@ -37,6 +37,7 @@ export function createImage (globalOptions: CreateImageOptions) {
   $img.getImage = getImage
   $img.getMeta = ((input: string, options?: ImageOptions) => getMeta(ctx, input, options)) as $Img['getMeta']
   $img.getSizes = ((input: string, options: ImageSizesOptions) => getSizes(ctx, input, options)) as $Img['getSizes']
+  $img.getDensitySet = ((input: string, options: ImageSizesOptions) => getDensitySet(ctx, input, options)) as $Img['getSrcSet']
```
Comment (hartmut-co-uk) [inline src/runtime/image.ts:40]:
```suggestion
  $img.getDensitySet = ((input: string, options: ImageSizesOptions) => getDensitySet(ctx, input, options)) as $Img['getDensitySet']
```

File: src/runtime/image.ts [hunk 6]
```diff
@@ -130,7 +131,9 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
   const width = parseSize(opts.modifiers?.width)
   const height = parseSize(opts.modifiers?.height)
   const hwRatio = (width && height) ? height / width : 0
-  const variants = []
+  const densities = opts.densities ? parseDensities(opts.densities) : ctx.options.densities
+  const sizeVariants = []
+  const srcVariants = []
```
Comment (hartmut-co-uk) [inline src/runtime/image.ts:136]:
```suggestion
  // const densities = opts.densities ? parseDensities(opts.densities) : ctx.options.densities
  const densities = opts.densities ? parseDensities(opts.densities) : (ctx.options.densities.length ? ctx.options.densities : [1, 2]) // TODO: pls help, default should come from `ctx.options.densities` ?!?
  const sizeVariants = []
  const srcVariants = []
```

File: src/runtime/image.ts [hunk 7]
```diff
@@ -163,25 +166,68 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
...
     })
+
+    if (densities) {
+      for (const density of densities) {
+        srcVariants.push({
+          width: _cWidth * density,
+          src: ctx.$img!(input, { ...opts.modifiers, width: _cWidth * density, height: _cHeight ? _cHeight * density : undefined }, opts)
+        })
+      }
+    }
```
Comment (hartmut-co-uk) [inline src/runtime/image.ts:184]:
```suggestion
    sizeVariants.push({
      size,
      screenMaxWidth,
      media: `(max-width: ${screenMaxWidth}px)`
    })

    // add variants for srcset for all densities
    if (densities) {
      for (const density of densities) {
        srcVariants.push({
          width: _cWidth * density,
          src: ctx.$img!(input, { ...opts.modifiers, width: _cWidth * density, height: _cHeight ? _cHeight * density : undefined }, opts)
        })
      }
    }
```

File: src/runtime/image.ts [hunk 8]
```diff
@@ -163,25 +166,68 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
...
+  let previousSize = ''
+  for (let i = sizeVariants.length - 1; i >= 0; i--) {
+    const sizeVariant = sizeVariants[i]
+    if (sizeVariant.media === previousSize) {
+      sizeVariants.splice(i, 1)
+    }
+    previousSize = sizeVariant.size
   }
 
+  srcVariants.sort((v1, v2) => v1.width - v2.width)
```
Comment (hartmut-co-uk) [inline src/runtime/image.ts:198]:
```suggestion
  // for last size, always remove media (convention)
  if (sizeVariants[sizeVariants.length - 1]) {
    sizeVariants[sizeVariants.length - 1].media = ''
  }

  // de-duplicate sizeVariants
  let previousMedia = null
  for (let i = sizeVariants.length - 1; i >= 0; i--) {
    const sizeVariant = sizeVariants[i]
    if (sizeVariant.media === previousMedia) {
      sizeVariants.splice(i, 1)
    }
    previousMedia = sizeVariant.media
  }

  srcVariants.sort((v1, v2) => v1.width - v2.width)

  // de-duplicate srcVariants
  let previousWidth = null
  for (let i = srcVariants.length - 1; i >= 0; i--) {
    const sizeVariant = srcVariants[i]
    if (sizeVariant.width === previousWidth) {
      srcVariants.splice(i, 1)
    }
    previousWidth = sizeVariant.width
  }
```

File: src/runtime/image.ts [hunk 9]
```diff
@@ -163,25 +166,68 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
...
-    srcset: variants.map(v => `${v.src} ${v.width}w`).join(', '),
+    sizes: sizeVariants.map(v => `${v.media ? v.media + ' ' : ''}${v.size}`).join(', '),
+    srcset: srcVariants.map(v => `${v.src} ${v.width}w`).join(', '),
     src: defaultVar?.src
   }
 }
+
+function getDensitySet (ctx: ImageCTX, input: string, opts: ImageSizesOptions): string|undefined {
+  const srcSet :{ density: string, src: string }[] = []
+  let densities = opts.densities ? parseDensities(opts.densities) : ctx.options.densities
```
Comment (hartmut-co-uk) [inline src/runtime/image.ts:211]:
```suggestion
  // let densities = opts.densities ? parseDensities(opts.densities) : ctx.options.densities
  let densities = opts.densities ? parseDensities(opts.densities) : (ctx.options.densities.length ? ctx.options.densities : [1, 2]) // TODO: pls help, default should come from `ctx.options.densities` ?!?
```

File: src/module.ts [hunk 10]
```diff
@@ -42,7 +43,8 @@ export default defineNuxtModule<ModuleOptions>({
     },
     internalUrl: '',
     providers: {},
-    alias: {}
+    alias: {},
+    densities: []
```
Comment (hartmut-co-uk) [inline src/module.ts:47]:
```suggestion
    densities: [1, 2]
```

**Review by hartmut-co-uk — CHANGES_REQUESTED** [review]
All proposed code changes should fix the implementation, I also have adjusted tests, not sure how to propose changes via PR suggestions when the files have not been changed yet..

**Comment by hartmut-co-uk** [comment]
also to fix existing tests, I have made following changes:

**`test/e2e/generate.test.ts`**
```ts
import { fileURLToPath } from 'node:url'

import { describe, it, expect } from 'vitest'
import { setup, useTestContext } from '@nuxt/test-utils'
import { useNuxt } from '@nuxt/kit'
import { resolve } from 'pathe'
import { globby } from 'globby'

await setup({
  rootDir: fileURLToPath(new URL('../../playground', import.meta.url)),
  build: true,
  nuxtConfig: {
    image: {
      inject: false
    },
    hooks: {
      'modules:before' () {
        const nuxt = useNuxt()
        nuxt.options.nitro.prerender = { routes: ['/provider/ipx'] }
      }
    }
  }
})

describe('ipx provider', () => {
  it('generates static files', async () => {
    const ctx = useTestContext()
    const outputDir = resolve(ctx.nuxt!.options.nitro.output?.dir || '', 'public/_ipx')
    const files = await globby(outputDir)
    expect(files.sort().map(f => f.replace(outputDir, '/_ipx'))).toMatchInlineSnapshot(`
      [
        "/_ipx/_/images/nuxt.png",
        "/_ipx/s_300x300/images/colors.jpg",
        "/_ipx/s_300x300/images/everest.jpg",
        "/_ipx/s_300x300/images/tacos.svg",
        "/_ipx/s_300x300/unsplash/photo-1606112219348-204d7d8b94ee",
        "/_ipx/s_600x600/images/colors.jpg",
        "/_ipx/s_600x600/images/everest.jpg",
        "/_ipx/s_600x600/images/tacos.svg",
        "/_ipx/s_600x600/unsplash/photo-1606112219348-204d7d8b94ee",
      ]
    `)
  })
})

```


**`test/unit/image.test.ts`**
```ts
// @vitest-environment nuxt

import { beforeEach, describe, it, expect } from 'vitest'
import { ComponentMountingOptions, VueWrapper, mount } from '@vue/test-utils'

import { NuxtImg } from '#components'

describe('Renders simple image', () => {
  let wrapper: VueWrapper<any>
  const src = '/image.png'

  beforeEach(() => {
    wrapper = mountImage({
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })
  })

  it('Matches snapshot', () => {
    expect(wrapper.html()).toMatchInlineSnapshot('"<img src=\\"/_ipx/s_1800x1800/image.png\\" width=\\"200\\" height=\\"200\\" data-nuxt-img=\\"\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/s_500x500/image.png 500w, /_ipx/s_900x900/image.png 900w, /_ipx/s_1000x1000/image.png 1000w, /_ipx/s_1800x1800/image.png 1800w\\">"')
  })

  it('props.src is picked up by getImage()', () => {
    const domSrc = wrapper.element.getAttribute('src')
    expect(domSrc).toMatchInlineSnapshot('"/_ipx/s_1800x1800/image.png"')
  })

  it('props.src is reactive', async () => {
    const newSource = '/image.jpeg'
    wrapper.setProps({ src: newSource })
    await nextTick()

    const domSrc = wrapper.find('img').element.getAttribute('src')
    expect(domSrc).toMatchInlineSnapshot('"/_ipx/s_1800x1800/image.jpeg"')
  })

  it('sizes', () => {
    const sizes = wrapper.find('img').element.getAttribute('sizes')
    expect(sizes).toBe('(max-width: 500px) 500px, 900px')
  })

  it('encodes characters', () => {
    const img = mountImage({
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src: '/汉字.png'
    })
    expect(img.html()).toMatchInlineSnapshot('"<img src=\\"/_ipx/s_1800x1800/%E6%B1%89%E5%AD%97.png\\" width=\\"200\\" height=\\"200\\" data-nuxt-img=\\"\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/s_500x500/%E6%B1%89%E5%AD%97.png 500w, /_ipx/s_900x900/%E6%B1%89%E5%AD%97.png 900w, /_ipx/s_1000x1000/%E6%B1%89%E5%AD%97.png 1000w, /_ipx/s_1800x1800/%E6%B1%89%E5%AD%97.png 1800w\\">"')
  })

  it('correctly sets crop', () => {
    const img = mountImage({
      src: '/image.png',
      width: 1000,
      height: 2000,
      sizes: 'xs:100vw sm:100vw md:300px lg:350px xl:350px 2xl:350px'
    })
    expect(img.html()).toMatchInlineSnapshot('"<img src=\\"/_ipx/s_1280x2560/image.png\\" width=\\"1000\\" height=\\"2000\\" data-nuxt-img=\\"\\" sizes=\\"(max-width: 320px) 100vw, (max-width: 640px) 100vw, (max-width: 768px) 300px, (max-width: 1024px) 350px, (max-width: 1280px) 350px, 350px\\" srcset=\\"/_ipx/s_300x600/image.png 300w, /_ipx/s_320x640/image.png 320w, /_ipx/s_350x700/image.png 350w, /_ipx/s_600x1200/image.png 600w, /_ipx/s_640x1280/image.png 640w, /_ipx/s_700x1400/image.png 700w, /_ipx/s_1280x2560/image.png 1280w\\">"')
  })
})

const mountImage = (props: ComponentMountingOptions<typeof NuxtImg>['props']) => mount(NuxtImg, { props })

```


**`test/unit/picture.test.ts`**
```ts
// @vitest-environment nuxt

import { beforeEach, describe, it, expect } from 'vitest'
import { VueWrapper, mount } from '@vue/test-utils'
import { NuxtPicture } from '#components'

describe('Renders simple image', () => {
  let wrapper: VueWrapper<any>
  const src = '/image.png'

  const observer = {
    wasAdded: false,
    wasDestroyed: false
  }

  beforeEach(() => {
    window.IntersectionObserver = class IntersectionObserver {
      root: any
      rootMargin: any
      thresholds: any
      takeRecords: any

      observe (_target: Element) {
        observer.wasAdded = true
      }

      disconnect () {
        observer.wasDestroyed = true
      }

      unobserve () {
        observer.wasDestroyed = true
      }
    }
    wrapper = mount(NuxtPicture, {
      propsData: {
        loading: 'lazy',
        width: 200,
        height: 200,
        sizes: '200,500:500,900:900',
        src
      }
    })
  })

  it('Matches snapshot', () => {
    expect(wrapper.html()).toMatchInlineSnapshot(`
      "<picture>
        <source type=\\"image/webp\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/f_webp&s_500x500/image.png 500w, /_ipx/f_webp&s_900x900/image.png 900w, /_ipx/f_webp&s_1000x1000/image.png 1000w, /_ipx/f_webp&s_1800x1800/image.png 1800w\\"><img width=\\"200\\" height=\\"200\\" data-nuxt-pic=\\"\\" src=\\"/_ipx/f_png&s_1800x1800/image.png\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/f_png&s_500x500/image.png 500w, /_ipx/f_png&s_900x900/image.png 900w, /_ipx/f_png&s_1000x1000/image.png 1000w, /_ipx/f_png&s_1800x1800/image.png 1800w\\">
      </picture>"
    `)
  })

  it.todo('alt attribute is generated')

  it('props.src is picked up by getImage()', () => {
    [['source', 'srcset', '/_ipx/f_webp&s_500x500/image.png'], ['img', 'src']].forEach(([element, attribute, customSrc]) => {
      const domSrc = wrapper.find(element).element.getAttribute(attribute)
      expect(domSrc).toContain(customSrc || src)
    })
  })

  it('renders webp image source', () => {
    expect(wrapper.find('[type="image/webp"]').exists()).toBe(true)
  })

  it('props.src is reactive', async () => {
    const newSource = '/image.jpeg'
    wrapper.setProps({ src: newSource })

    await nextTick()

    ;[['source', 'srcset', '/_ipx/f_webp&s_500x500/image.jpeg'], ['img', 'src']].forEach(([element, attribute, src]) => {
      const domSrc = wrapper.find(element).element.getAttribute(attribute)
      expect(domSrc).toContain(src || newSource)
    })
  })

  it('sizes', () => {
    const sizes = wrapper.find('source').element.getAttribute('sizes')
    expect(sizes).toBe('(max-width: 500px) 500px, 900px')
  })

  it('renders src when svg is passed', () => {
    const wrapper = mount(NuxtPicture, {
      propsData: {
        src: '/image.svg'
      }
    })
    expect(wrapper.html()).toMatchInlineSnapshot('"<picture><img data-nuxt-pic=\\"\\" src=\\"/image.svg\\"></picture>"')
  })

  it('encodes characters', () => {
    const img = mount(NuxtPicture, {
      propsData: {
        loading: 'lazy',
        width: 200,
        height: 200,
        sizes: '200,500:500,900:900',
        src: '/汉字.png'
      }
    })
    expect(img.html()).toMatchInlineSnapshot(`
      "<picture>
        <source type=\\"image/webp\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/f_webp&s_500x500/%E6%B1%89%E5%AD%97.png 500w, /_ipx/f_webp&s_900x900/%E6%B1%89%E5%AD%97.png 900w, /_ipx/f_webp&s_1000x1000/%E6%B1%89%E5%AD%97.png 1000w, /_ipx/f_webp&s_1800x1800/%E6%B1%89%E5%AD%97.png 1800w\\"><img width=\\"200\\" height=\\"200\\" data-nuxt-pic=\\"\\" src=\\"/_ipx/f_png&s_1800x1800/%E6%B1%89%E5%AD%97.png\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_ipx/f_png&s_500x500/%E6%B1%89%E5%AD%97.png 500w, /_ipx/f_png&s_900x900/%E6%B1%89%E5%AD%97.png 900w, /_ipx/f_png&s_1000x1000/%E6%B1%89%E5%AD%97.png 1000w, /_ipx/f_png&s_1800x1800/%E6%B1%89%E5%AD%97.png 1800w\\">
      </picture>"
    `)
  })
})

```

File: src/runtime/providers/directus.ts [hunk 11]
```diff
@@ -8,8 +8,10 @@ export const operationsGenerator = createOperationsGenerator({
 
 export const getImage: ProviderGetImage = (src, { modifiers = {}, baseURL } = {}) => {
   // Separating the transforms from the rest of the modifiers
-  const transforms = modifiers.transforms
+  let transforms = modifiers.transforms
   if (transforms && transforms.length > 0) {
+    // de-duplicate (can get multiplied when haveing >1 densities configured)
+    transforms = transforms.filter((value: any, index: any) => transforms.indexOf(value) === index)
```
Comment (danielroe) [inline src/runtime/providers/directus.ts:14]: I think we might need to try JSON.stringify on the values to be compared as I think we're not always guaranteed they are the _same_ array.

File: docs/content/5.configuration.md [hunk 12]
```diff
@@ -154,6 +154,20 @@ export default defineNuxtConfig({
 })
 ```
 
+## `densities`
+
+Default: `[1, 2]`
+
+Specify a value to work with DevicePixelRation > 1 (these are devices with Retina display and others). You must specify for which DevicePixelRation values you want to adapt images.
```
Comment (danielroe) [inline docs/content/5.configuration.md:161]:
```suggestion
Specify a value to work with `devicePixelRatio` > 1 (these are devices with retina display and others). You must specify for which `devicePixelRatio` values you want to adapt images.
```

File: src/runtime/components/nuxt-img.ts [hunk 13]
```diff
@@ -41,6 +42,16 @@ export default defineComponent({
       if (props.sizes && (!props.placeholder || placeholderLoaded.value)) {
         attrs.sizes = sizes.value.sizes
         attrs.srcset = sizes.value.srcset
+      } else if (props.densities || $img.options.densities) {
+        attrs.srcset = $img.getDensitySet(props.src!, {
+          ..._base.options.value,
+          densities: props.densities,
```
Comment (danielroe) [inline src/runtime/components/nuxt-img.ts:48]: should this be:
```suggestion
          densities: props.densities || $img.options.densities,
```

Comment (hartmut-co-uk) [reply src/runtime/components/nuxt-img.ts:48]: I think this should already be handled here?
https://github.com/nuxt/image/pull/769/files#diff-6b00e0c4d4cf58d68124cc022436a52ea9dd527851ad8a1397cfa336b2975a4fR229

File: src/runtime/image.ts [hunk 14]
```diff
@@ -163,25 +167,85 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
...
   }
 }
+
+function getDensitySet (ctx: ImageCTX, input: string, opts: ImageSizesOptions): string|undefined {
+  const densities = opts.densities ? parseDensities(opts.densities) : ctx.options.densities
+  if (densities.length === 0) {
+    return undefined
+  }
+
+  const srcsetVariants :{ density: string, src: string }[] = []
```
Comment (danielroe) [inline src/runtime/image.ts:234]:
```suggestion
  const srcsetVariants: Array<{ density: string, src: string }> = []
```

File: src/runtime/image.ts [hunk 15]
```diff
@@ -163,25 +167,85 @@ function getSizes (ctx: ImageCTX, input: string, opts: ImageSizesOptions) {
...
   return {
-    sizes: variants.map(v => `${v.media ? v.media + ' ' : ''}${v.size}`).join(', '),
-    srcset: variants.map(v => `${v.src} ${v.width}w`).join(', '),
+    sizes: sizeVariants.map(v => `${v.media ? v.media + ' ' : ''}${v.size}`).join(', '),
+    srcset: srcsetVariants.map(v => `${v.src} ${v.width}w`).join(', '),
     src: defaultVar?.src
   }
 }
+
+function getDensitySet (ctx: ImageCTX, input: string, opts: ImageSizesOptions): string|undefined {
```
Comment (danielroe) [inline src/runtime/image.ts:228]:
```suggestion
function getDensitySet (ctx: ImageCTX, input: string, opts: ImageSizesOptions): string | undefined {
```

**Review by hartmut-co-uk — APPROVED** [review]

**Review by danielroe — APPROVED** [review]
Thank you all for amazing work on this.

**Pull request merged**